lib/krb5/principal.c

```diff
@@ -205,200 +206,197 @@ krb5_principal_get_num_comp(krb5_context context,
 	krb5_set_error_message(context, KRB5_ERR_NO_SERVICE,
 			       N_("Can't require both realm and "
 				  "no realm at the same time", ""));
 	return KRB5_ERR_NO_SERVICE;
     }
 
-    /* count number of component,
-     * enterprise names only have one component
-     */
-    ncomp = 1;
-    if (!enterprise) {
-	for (p = name; *p; p++) {
-	    if (*p=='\\') {
-		if (!p[1]) {
-		    krb5_set_error_message(context, KRB5_PARSE_MALFORMED,
-					   N_("trailing \\ in principal name", ""));
-		    return KRB5_PARSE_MALFORMED;
-		}
-		p++;
-	    } else if (*p == '/')
-		ncomp++;
-	    else if (*p == '@')
-		break;
-	}
-    }
-    comp = calloc(ncomp, sizeof(*comp));
-    if (comp == NULL)
-	return krb5_enomem(context);
+    *out = calloc(1, sizeof(**out));
 
-    n = 0;
-    p = start = q = s = strdup(name);
-    if (start == NULL) {
-	free(comp);
-	return krb5_enomem(context);
+    in = strdup(name);
+    if (!in) {
+	ret = krb5_enomem(context);
+	goto out;
     }
-    while (*p) {
-	c = *p++;
-	if (c == '\\') {
-	    c = *p++;
-	    if (c == 'n')
-		c = '\n';
-	    else if (c == 't')
-		c = '\t';
-	    else if (c == 'b')
-		c = '\b';
-	    else if (c == '0') {
-                /*
-                 * We'll ignore trailing embedded NULs in components and
-                 * realms, but can't support any other embedded NULs.
-                 */
-                while (*p) {
-                    if ((*p == '/' || *p == '@') && !got_realm)
-                        break;
-                    if (*(p++) != '\\' || *(p++) != '0') {
-                        ret = KRB5_PARSE_MALFORMED;
-                        krb5_set_error_message(context, ret,
-                                               N_("embedded NULs in principal "
-                                                  "name not supported", ""));
-                        goto exit;
-                    }
-                }
-                continue;
-            } else if (c == '\0') {
-		ret = KRB5_PARSE_MALFORMED;
-		krb5_set_error_message(context, ret,
-				       N_("trailing \\ in principal name", ""));
-		goto exit;
-	    }
-	} else if (enterprise && first_at) {
-	    if (c == '@')
-		first_at = 0;
-	} else if ((c == '/' && !enterprise) || c == '@') {
-	    if (got_realm) {
-		ret = KRB5_PARSE_MALFORMED;
-		krb5_set_error_message(context, ret,
-				       N_("part after realm in principal name", ""));
-		goto exit;
-	    } else {
-		comp[n] = malloc(q - start + 1);
-		if (comp[n] == NULL) {
-		    ret = krb5_enomem(context);
-		    goto exit;
+
+    for (pos=in, i=0, j=0; in[i]; i++) {
+	switch (UNVIS(&outc, in[i], &visstate, 0)) {
+	case 0:
+	case UNVIS_NOCHAR:
+	    escaped = 1;
+	    break;
+	case UNVIS_VALID:
+	    if (!escaped && ((!enterprise && outc == '/') || outc == '@')) {
+		if (enterprise && (state & FIRST_AT)) {
+		    state &= ~FIRST_AT;
+		} else {
+		    ret = add_bit(context, out, &state, outc, pos, &in[j]);
+		    if (ret)
+			goto out;
+		    pos = &in[i];
```

Comment:
It must be that unvis(3) on MacOS does not work properly. This passed the tests on NetBSD, travis, and appveyor. There is an implementation of unvis(3) in libroken, perhaps we need to specify that it be used on OSes with a deficient unvis(3). I am using unvis(3) here, specifically to allow for additional escaping in the form of octal and a few things that the original code did not support because, as we discussed and, if you recall, you suggested, we want to escape all unprintable characters.
I can comment up the code a little bit, but it is a relatively standard use of unvis(3) according to the man page. My only innovations here are:
    1. the UNVIS() macro which fixes "\ " and improves performance a bit, and
    2. keeping track of whether we are escaped so that I can correctly process unescaped
        "/" and "@".

Reply:
> It must be that unvis(3) on MacOS does not work properly.

We could add some configure tests to determine whether the platform's `unvis(3)` meets our requirements. What does/should `unvis()` do with `\@, \/,` or generally other characters that don't need escaping, but are escaped anyway?

Note that I also added `*(ST) = 0`, when we're done unescaping SPACE, AT, FORWARD_SLASH outside `unvis()`, because otherwise unvis may erroneously believe the next character to still be escaped, e.g. `foo\ \n` will decode to 'foo \n' not `foo <NEWLINE>`.

Reply:
Below is a test of escaping all non-control non-alnum characters on MacOS with unvis(1) (and then `cat -v`). The output shows that generally `\<special>` fails to parse, with two exceptions:
- \$ is silently discarded
- `\^` turns the __next__ character into a control, e.g. `\^b` is `\002`
```text
in: foo\ bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\!bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\"bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\#bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\$bar
out: foobar
in: foo\%bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\&bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\'bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\(bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\)bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\*bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\+bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\,bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\-bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\.bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\/bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\:bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\;bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\<bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\=bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\>bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\?bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\@bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\[bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\]bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\^bar
out: foo^Bar
in: foo\_bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\`bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\{bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\|bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\}bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
in: foo\~bar
unvis: <stdin>: offset: 5: can't decode
out: foobar
```

Reply:
Yeah, here is my output:
```
$ unvis
foo\@bar
foo@bar
foo\/bar
foo/bar
foo\ bar
unvis: <stdin>: offset: 23: can't decode
```
So, it is a bit different.

Reply:
Thinking about it, it may be better to put / and @ in the macro than to add to the burden of autoconf. I think that MacOS must have imported an older version of `unvis(3)` and we should, of course, support it. I was just explaining why I did not see this issue on my setup. I'll adopt your patch on mine. Good spot on the *(ST) = 0.

Reply:
One thing I'm not clear about is whether we need to be able to handle any of the other specials. That is, what should this code do with `foo\^bar@MIT.EDU` or `foo\$bar@MIT.EDU`?  Is that valid input, or will we never see that? And similar for the other escaped specials that MacOS `unvis()` believes to be bad syntax...

Reply:
If those are out of scope, yes, just adding `@` and `\` may be enough.

Reply:
I'm using unvis(3) because I want to be able to handle other escape sequences, most specifically, I've added octal encoding to `krb5_unparse_name()`. IMO, foo\^bar@MIT.EDU should be currently undefined and so it's fine to add a definition of it.

Reply:
I've taken the krb5_{,un}parse_name() patches out of this pull request and opened:
Improve encoding of krb5_{,un}parse_principal(). #658